File: lab3KononKI305/src/SmartPhone.java
// SmartPhone.java
import java.util.Scanner;

public class SmartPhone extends Phone implements InternetConnectable {
    private boolean internetConnected;

    // Default constructor
    public SmartPhone() {
        super();
        internetConnected = false;
    }

    // Constructor with parameters
    public SmartPhone(String brand, String model) {
        super(brand, model);
        internetConnected = false;
    }

    // Implementing the InternetConnectable interface
    @Override
    public void connectToInternet() {
        if (isPowerOn()) {
            System.out.println("Connecting to the Internet...");
            internetConnected = true;
        } else {
            System.out.println("Cannot connect to the Internet. Phone is powered off.");
        }
    }

    // Override the makeCall method to include internet connection check
    @Override
    public void makeCall(String phoneNumber) {
        if (internetConnected) {
            super.makeCall(phoneNumber);
        } else {
            System.out.println("Cannot make a call. No Internet connection.");
        }
    }

    // Override the sendMessage method to include internet connection check
    @Override
    public void sendMessage(String phoneNumber, String message) {
        if (internetConnected) {
            super.sendMessage(phoneNumber, message);
        } else {
            System.out.println("Cannot send a message. No Internet connection.");
        }
    }

    // Additional methods specific to SmartPhone can be added here

    // Getter for internetConnected
    public boolean isInternetConnected() {
        return internetConnected;
    }
    @Override
    public boolean supportsInternet() {
        return true;
    }
}

File: lab9KononKI305/smart_phone.py
# smart_phone.py
from phone import Phone
from internet_connectable import InternetConnectable
from call_history import CallHistory

class SmartPhone(Phone):


    def __init__(self, brand="Unknown", model="Unknown"):
        super().__init__(brand, model)
        self.internet_connected = False

    def connect_to_internet(self):
        if self.is_turned_on():
            print("Connecting to the Internet...")
            self.internet_connected = True
        else:
            print("Cannot connect to the Internet. Phone is powered off.")

    def make_call(self, phone_number):
        if self.internet_connected:
            super().make_call(phone_number)
        else:
            print("Cannot make a call. No Internet connection.")

    def send_message(self, phone_number, message):
        if self.internet_connected:
            super().send_message(phone_number, message)
        else:
            print("Cannot send a message. No Internet connection.")

    def supports_internet(self):
        return True